akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala

```diff
@@ -0,0 +1,127 @@
+        } else {
+          HttpResponse(status = StatusCodes.MovedPermanently, headers = List(Location(s"/r${c + 1}")))
+        }
+      }, serverHostName, serverPort)
+
+      val N = 100
+      val result = Source.fromIterator(() ⇒ Iterator.from(1))
+        .take(N)
+        .map(id ⇒ HttpRequest(uri = s"/r$id"))
+        .via(Http().outgoingConnection(serverHostName, serverPort))
+        .mapAsync(4)(_.entity.toStrict(1.second))
+        .map { r ⇒ val s = r.data.utf8String; log.debug(s); s.toInt }
+        .runFold(0)(_ + _)
+
+      result.futureValue(PatienceConfig(10.seconds)) shouldEqual (N + 2) * (N / 2) // (2*2) + (4*2) + ... + (100 * 2)
+      binding.futureValue.unbind()
+
+    }
+
+    "be able to handle infinite redirect loop" in Utils.assertAllStagesStopped {
+      val (_, serverHostName, serverPort) = TestUtils.temporaryServerHostnameAndPort()
+      val cycleLength = 3
+      val binding = Http().bindAndHandleSync(r ⇒ {
+        val c = r.uri.toString.reverse.takeWhile(Character.isDigit).reverse.toInt
+        HttpResponse(status = StatusCodes.MovedPermanently, headers = List(Location(s"/r${(c % cycleLength) + 1}")))
+      }, serverHostName, serverPort)
+
+      val x = Source.fromIterator(() ⇒ Iterator.from(1))
+        .take(cycleLength)
+        .map(id ⇒ HttpRequest(uri = s"/r$id"))
+        .via(Http().outgoingConnection(serverHostName, serverPort))
+        .runWith(Sink.head)
+
+      a[RedirectSupportStage.InfiniteRedirectLoopException.type] should be thrownBy Await.result(x, 3.second)
+
+      binding.futureValue.unbind()
+    }
+
+    "be able to forward selected headers to redirected requests in same origin" in Utils.assertAllStagesStopped {
+      val (_, serverHostName, serverPort) = TestUtils.temporaryServerHostnameAndPort()
+      val binding = Http().bindAndHandleSync(r ⇒ {
+        val c = r.uri.toString.reverse.takeWhile(Character.isDigit).reverse.toInt
+        if (c % 2 == 0) {
+          HttpResponse(headers = List(headers.ETag(r.headers.find(_.is("id")).map(_.value).getOrElse("0"))))
+        } else {
+          HttpResponse(status = StatusCodes.MovedPermanently, headers = List(Location(s"/r${c + 1}")))
+        }
+      }, serverHostName, serverPort)
+
+      val N = 100
+      val result = Source.fromIterator(() ⇒ Iterator.from(1))
+        .take(N)
+        .map(id ⇒ HttpRequest(uri = s"/r$id", headers = List(RawHeader("id", id.toString))))
+        .via(Http().outgoingConnection(serverHostName, serverPort))
+        .map(p ⇒ {
+          if (p.status.isRedirection) throw new Exception("No redirects here!")
+          p.headers.find(_.is("etag")).map(_.value.replace("\"", "").toInt).getOrElse(0)
+        })
+        .runFold(0)(_ + _)
+
+      result.futureValue(PatienceConfig(1000.seconds)) shouldEqual (N + 1) * (N / 2)
+
+      binding.futureValue.unbind()
+    }
+
+    "be able to forward selected headers to redirected requests in different origin" /*in Utils.assertAllStagesStopped */ ignore {
```

Comment:
I'll check out the branch and investigate this one then; likely not this week though – easter happened.
Sorry for the delay!